src/proposal.mligo

```diff
@@ -337,6 +391,39 @@ let delete_proposal
     Set.remove (start_date, proposal_key) store.proposal_key_list_sort_by_date
   }
 
+let update_quorum(store, config : storage * config): storage =
+  let current_period = get_current_period_num(store.start_time, config.voting_period) in
+  if store.quorum_threshold_at_stage.last_updated_period = current_period
+    then store // Quorum has been updated in this period, so no change is required.
+    else
+      let last_period_frozen_tokens = match Map.find_opt 0n store.total_supply with
+        | Some (n) -> n
+            // We assume that this function will always be called before freezing
+            // tokens, so that if we are in this branch (where quorum_threshold_at_stage
+            // is outdated), we assume the total supply will be outdated as well and will
+            // have frozen tokens available in last/earlier period.
+        | None -> ((failwith "FROZEN_TOTAL_SUPPLY_NOT_FOUND") : nat) in
```

Comment:
well yes, but if we got here we are already in a catastrophically inconsistent state...

The idea is to keep the contract as simple as possible (to save on execution/size costs) and prove with tests that (in this case) none of the functions that modify the `total_supply` can remove this entry.

Reply:
I think it is, the only case in which this could happen is:
1. the contract was originated with an empty `total_supply` in which case the behavior of `0n` is correct.
2. the contract has a serious bug and as a result there is no `total_supply` of frozen tokens

I mean this is an error that we cannot even test for, unless we purposefully originate the contract without a `total_supply` entry, in which case as I said is ok to behave like 0.

In case of (2) the contract:
- if with an error is stuck: as there is no way to `freeze`, `unfreeze`, `flush` or `drop_proposal`
- without an error produces a 0% `previous_participation`, which results in a lower quorum threshold

I agree that in general this is something to raise an error for, but this is one case where the disadvantage can be ignored, as the problem that caused it (no `total_supply`) has a far larger impact.
In other word, all the entrypoints "affected" by this returning `0n` don't work without a frozen `total_supply` anyhow:
- `freeze` will throw `FA2_TOKEN_UNDEFINED` in `credit_to`
- `unfreeze` will do the same in `debit_from`
- `flush` will also fail, because even in config proposals and such it has to give back/consume/burn tokens 
- `drop_proposal` will also throw `FA2_TOKEN_UNDEFINED` for either `credit_to` or `debit_from`

whatever this returns in such a case is in-influential, as the contract is utterly broken either way.

Reply:
First of all, I think it was a bad idea to re-purpose `total_supply` to indicate the support for various token ids. I think that is the root cause that we find ourself with disagreement here. Because in normal case `total_supply` to be zero would be perfectly fine. 

The problem with your reasoning that I see is this. Once we find ourselves in an inconsistent state at some point in future, any answer to the question as to how we ended up in that state, can only be speculative. Because assumptions that you make right now about what the rest of the code is doing, might not hold at that point in future.

Now, to take this discussion a bit in a different direction, I am not sure what problem you find with throwing error, instead of returning zero. You say that we want to keep the contract simple, but AFAIU, throwing this error does not add to its complexity in any way.

Reply:
> I think it was a bad idea to re-purpose total_supply to indicate the support for various token ids.
> in normal case total_supply to be zero would be perfectly fine.

That's true, however it's a very simple way to support this, a more clear alternative would have been to add a field in storage just for the set of supported `token_id`, but that's an added cost that we are so avoiding.

> any answer to the question as to how we ended up in that state, can only be speculative

That's fair, but we know for sure that it wasn't intended and my reasoning is based on the results of such a situation, not the cause. What happens to the entrypoints' logic is true regardless of how we got there.

> I am not sure what problem you find with throwing error, instead of returning zero.
> throwing this error does not add to its complexity in any way.

Well I don't have a problem with it really, we can go with it and as I said it's the proper thing to do.
My only motivations are that:
1. it adds cost, as `PUSH 0` is less expensive then `PUSH "FROZEN_TOTAL_SUPPLY_NOT_FOUND"; FAILWITH`.
    (both to originate and run)
2. it doesn't add any more safety, as the entrypoints affected could not be executed in that case regardless

Reply:
Your points are valid tho, IMO we can leave this for later as the discussion is more based on our use of the `total_supply` to indicate the support for tokens and a slight difference in efficiency.

We can go with your `FAILWITH` approach and get back to this discussion if we need to cut the cost down (or reconsider the `total_supply` field use), I'm ok with that :+1: 
If you do so however, do the same in `do_total_vote_meet_quorum_threshold` for consistency and remember to document this error in the spec as well, ok?

Reply:
> it doesn't add any more safety, as the entrypoints affected could not be executed in that case regardless

Just want to clarify. 

AFAIU you are talking about a case where the execution start from a corrupt state. But since we pass around storage through various funcions, and  if one of them modify storage to have this inconsistent state, and throwing the exception as soon as we come across it might save ourselfves from writing down an inconsistent version of the storage, permanantly to the contract storage.

So in that regard,  I think it's more safe, because it allows us to recover storage from a valid state, rather than a corrupted one. 

>If you do so however, do the same in do_total_vote_meet_quorum_threshold for consistency and remember to document this error in the spec as well, ok?

Will do.

Reply:
Have added the error throwing to `do_total_vote_meet_quorum_threshold` function and updated spec. I have renamed the error to just 'BAD_STATE', as a generic error.

Reply:
> throwing the exception as soon as we come across it might save ourselfves from writing down an inconsistent version of the storage

Well keeping in mind that the storage doesn't actually change if anything in the whole operation fails this is less of a concern here, where we one of the failing functions is called in the same entrypoint.
However that's a valid point, it would be difficult to verify that this is the case and keeps being the case in the future. :ok_hand: